### PR TITLE
Allow Custom Headers to be managed via the API

### PR DESCRIPTION
**Pull request by griffri**
Resolves https://github.com/dlcs/protagonist/issues/539

### CODE REVIEW

File: src/protagonist/API.Tests/Integration/CustomHeaderTests.cs [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+            Customer = customerId,
+            Key = "test-key",
+            Value = "test-value"
+        };
+        
+        await dlcsContext.CustomHeaders.AddAsync(customHeader);
+        await dlcsContext.SaveChangesAsync();
+        
+        // Act
+        var response = await httpClient.AsCustomer(customerId).DeleteAsync(Path.Combine(path, customHeader.Id));
```
Comment (donaldgray) [inline src/protagonist/API.Tests/Integration/CustomHeaderTests.cs:89]: I think it's cleaner to have the entire path evident in the `path` variable, rather than building it where `httpClient` is called, that way at a quick scan of the tests it's really evident what URL is being tested.

Initially I thought this was testing `DELETE /customers{customerId}/customHeaders` and didn't initially clock the Id was at the end.

e.g.

```cs
[Fact]
public async Task Delete_CustomHeader_204()
{
    // Arrange
    const int customerId = 92;
    var path = $"customers/{customerId}/customHeaders/{customHeader.Id}`

    // other stuff

    var response = await httpClient.AsCustomer(customerId).DeleteAsync(path);
```

```cs
[Fact]
public async Task Delete_CustomHeader_204()
{
    // Arrange
    const int customerId = 92;
    var path = $"customers/{customerId}/customHeaders";

    // other stuff
    var response = await httpClient.AsCustomer(customerId).DeleteAsync(Path.Combine(path, customHeader.Id));
```

File: src/protagonist/API.Tests/Integration/CustomHeaderTests.cs [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+        
+        // Assert
+        response.StatusCode.Should().Be(HttpStatusCode.OK);
+        
+        var model = await response.ReadAsHydraResponseAsync<HydraCollection<CustomHeader>>();
+        model.Members.Should().HaveCount(2);
+    }
+    
+    [Fact]
+    public async Task Post_CustomHeader_201()
```
Comment (JackLewis-digirati) [inline src/protagonist/API.Tests/Integration/CustomHeaderTests.cs:120]: Do you need tests around the hydra validations? I.e. testing failure - same with what happens when you try and add a duplicate header?  Additionally, what happens when you try and post to a space which doesn't exist?

File: src/protagonist/API/Features/CustomHeaders/Requests/DeleteCustomHeader.cs [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    private readonly DlcsContext dbContext;
+
+    public DeleteCustomHeaderHandler(DlcsContext dbContext)
+    {
+        this.dbContext = dbContext;
+    }
+
+    public async Task<ResultMessage<DeleteResult>> Handle(DeleteCustomHeader request, CancellationToken cancellationToken)
+    {
+        var deleteResult = DeleteResult.NotFound;
```
Comment (donaldgray) [inline src/protagonist/API/Features/CustomHeaders/Requests/DeleteCustomHeader.cs:34]: `deleteResult` only ever has 2 different values, is it cleaner to use use the enum value directly rather than having a mutating variable? Using a specific enum value at the place it's used saves looking up in the code to see the current value.

E.g., in these it's evident what the return value is:
```cs
if (customHeader == null) return new ResultMessage<DeleteResult>("CustomHeader not found", DeleteResult.NotFound);

return new ResultMessage<DeleteResult>("CustomHeader deleted", DeleteResult.Deleted);
```

whereas here we need to look around to see what the current value of `deleteResult` is
```cs
if (customHeader == null) return new ResultMessage<DeleteResult>(message, deleteResult);

return new ResultMessage<DeleteResult>(message, deleteResult);
```

File: src/protagonist/API.Tests/Integration/CustomHeaderTests.cs [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+    [Fact]
+    public async Task Put_CustomHeader_404_IfNotFound()
+    {
+        // Arrange
+        const int customerId = 96;
+        var path = $"customers/{customerId}/customHeaders/{Guid.Empty}";
+        const string updatedCustomHeaderJson = @"{
+          ""key"": ""test-key-2"",
+          ""value"": ""test-value-2"",
+          ""space"": 2,
```
Comment (JackLewis-digirati) [inline src/protagonist/API.Tests/Integration/CustomHeaderTests.cs:190]: what happens when you set this value to a space which doesn't exist? Likely also needs verification that nothing changes on failure

File: src/protagonist/API/Features/CustomHeaders/Converters/CustomHeaderConverter.cs [hunk 5]
```diff
@@ -0,0 +1,28 @@
+﻿namespace API.Features.CustomHeaders.Converters;
+
+public static class CustomHeaderConverter
```
Comment (JackLewis-digirati) [inline src/protagonist/API/Features/CustomHeaders/Converters/CustomHeaderConverter.cs:3]: this class needs comments

File: src/protagonist/API.Tests/Features/CustomHeaders/Validation/HydraCustomHeaderValidator.cs [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+        var result = sut.TestValidate(customHeader);
+        result.ShouldHaveValidationErrorFor(ch => ch.Key);
+    }
+    
+    [Fact]
+    public void NewCustomHeader_Requires_Value()
+    {
+        var customHeader = new CustomHeader();
+        var result = sut.TestValidate(customHeader);
+        result.ShouldHaveValidationErrorFor(ch => ch.Value);
```
Comment (JackLewis-digirati) [inline src/protagonist/API.Tests/Features/CustomHeaders/Validation/HydraCustomHeaderValidator.cs:53]: might be worth validating that the values you haven't set are null in this set of tests?

**Review by donaldgray — APPROVED** [review]
Added a couple of comments.

If you change `"Related to"` in comment to one of the GH linking words (https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) it'll auto-link the issue and close it when you close this.

File: src/protagonist/API.Tests/Integration/CustomHeaderTests.cs (hunk 1 above)
Comment (griffri) [reply src/protagonist/API.Tests/Integration/CustomHeaderTests.cs]: Agreed. This has now been resolved in 17a80f5a6d1196fde444c6c9a291829124344379

File: src/protagonist/API/Features/CustomHeaders/Requests/DeleteCustomHeader.cs [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+        var message = string.Empty;
```
[the remark this answers is not in the document]
Comment (griffri) [reply src/protagonist/API/Features/CustomHeaders/Requests/DeleteCustomHeader.cs:35]: This should return a message if it cannot find a name query to delete - resolved in 01ce2bfb5832e9ea6641bb84dadc0d8725d65fba

File: src/protagonist/API/Features/CustomHeaders/Requests/DeleteCustomHeader.cs (hunk 3 above)
Comment (griffri) [reply src/protagonist/API/Features/CustomHeaders/Requests/DeleteCustomHeader.cs]: That's a good point, I've made it a little hard to follow. This has been changed in 95770330c2a1a080c057d35aab57ae96ae800260

File: src/protagonist/API.Tests/Integration/CustomHeaderTests.cs (hunk 4 above)
Comment (griffri) [reply src/protagonist/API.Tests/Integration/CustomHeaderTests.cs:190]: https://github.com/dlcs/protagonist/pull/569#discussion_r1280842836

File: src/protagonist/API/Features/CustomHeaders/Converters/CustomHeaderConverter.cs (hunk 5 above)
Comment (griffri) [reply src/protagonist/API/Features/CustomHeaders/Converters/CustomHeaderConverter.cs:3]: Good call - I missed the fact that the other converters have comments. I'll get to work on that 👍

File: src/protagonist/API.Tests/Features/CustomHeaders/Validation/HydraCustomHeaderValidator.cs (hunk 6 above)
Comment (griffri) [reply src/protagonist/API.Tests/Features/CustomHeaders/Validation/HydraCustomHeaderValidator.cs:59]: Good idea. Added in df51740576117eeabff930b0a854996b3ed7c0aa

File: src/protagonist/API.Tests/Integration/CustomHeaderTests.cs (hunk 2 above)
Comment (JackLewis-digirati) [reply src/protagonist/API.Tests/Integration/CustomHeaderTests.cs:120]: I mainly mean within the controller itself and while the tests themselves are probably not that useful, if you were to output a code coverage report, it would show
```csharp
if (!validationResult.IsValid)
{
    return this.ValidationFailed(validationResult);
}
```
is not covered

In terms of keys, is it worth writing a test to show you can push multiple keys in with the same id? Mainly because if I see key and value, I'd assume it's a dictionary which would mean the keys are unique - having a test like this just makes it cleared that the logic is intentional

And if you're allowing multiple of the same keys, is there a verified order that a key is returned in `GetCustomHeader`? I.e.: is it FIFO or LIFO? and is it worth writing a test to show this?

Comment (griffri) [reply src/protagonist/API.Tests/Integration/CustomHeaderTests.cs:120]: Tests for validation have been added in ddc6b699f722ac3d85cc181ef0fa1930f19baf2b

Comment (donaldgray) [reply src/protagonist/API.Tests/Integration/CustomHeaderTests.cs:120]: > And if you're allowing multiple of the same keys, is there a verified order that a key is returned in GetCustomHeader? I.e.: is it FIFO or LIFO? and is it worth writing a test to show this?

There is an order of precedence in how the keys are applied in orchestrator but nothing specified about how they should be returned here. Leaving it to DB order is fine for now.

Comment (griffri) [reply src/protagonist/API.Tests/Integration/CustomHeaderTests.cs:121]: a2bba53e489d958f5fa4d980f16cf2ad850df51b adds a test for multiple custom headers using the same ID
